design/one-pager-merge-jets-proposal.md

```diff
@@ -0,0 +1,402 @@
+* The single provider will be the classic one and we'll add Terrajet CRDs on
+  top.
+
+The order of execution will be roughly as following, taking
+[`provider-aws`][provider-aws] as an example:
+1. Add Terrajet pipeline to the provider repository.
+1. Copy the manual resource configurations from provider-jet-aws and introduce
+  the CRDs of those configurations in their new API group, without changing
+  their version, i.e. `v1alpha2`.
+1. Generate the full set of CRDs in `release-x.x-preview` branch similar to jet
+  repository and include those non-configured ones only in `-preview` suffixed
+  version.
+1. Write up a guide or tool to help the current users of non-preview version of
+  Jet providers to migrate to [`provider-aws`][provider-aws].
+
+We will also update the guidelines about how to add a resource roughly with the
+following:
+* If you are adding a new resource, use Terrajet pipeline unless there is a
+  blocker.
+* If you encounter a bug that cannot be solved with Terrajet in a meaningful
+  timeframe, consider changing its controller to classic hand-written controller
+  and bump the API version.
+* If an existing classic CRD (hand-written or ACK-generated) has a bug that
+  cannot be fixed easily or has a blocker or we continuously get bugs which
+  makes it hard to maintain, consider replacing it with Terrajet.
+
+The main idea behind these guidelines is that we'd like to make use of the best
+tooling for a given resource in terms of its maintenance cost and stability. We
+might end up with more and more classically written resources converted to
+Terrajet because it's easier to maintain but there is no goal of converting all
+existing stable resources, like VPC for example. For all sorts of conversions,
+we will make use of conversion webhooks support of Kubernetes API.
+
+There are several reasons to choose this approach instead of having providers
+separated by whether they call APIs directly or use Terrajet. Let's take a look
+at them one by one.
+
+### Dependence on Terraform Provider
+
+The generic Terrajet controller calls Terraform CLI which in turn calls its
+Terraform provider to operate. Each CRD has its own completely independent
+reconciler thread configured to work with the schema of that resource.
+
+This dependence causes the Crossplane provider to be exposed to the bugs of the
+underlying Terraform provider. While we strive to fix the bug there first, it's
+not always feasible for us to land that change in a timely manner. In addition,
+there are cases where we are locked to a version with a bug where downgrading or
+upgrading is not an option because of another bug or behavior change.
+
+With the mixed provider, we can change the underlying implementation to a
+hand-written controller whenever we face such a bug. With the Kubernetes API
+conversion tooling, we can manage the possible API changes as well. This would
+also mean that over time, we'd have less and less Terraform dependence in our
+ecosystem.
```

Comment:
Does a hand-written controller mean calling API/SDK to manage a cloud resource or its operation?

Reply:
For AWS, it specifically means calling API/SDK directly and without using ACK codegen pipeline. For GCP and Azure, there is no such pipeline anyway, so it's just classic implementation.